src/nvim/lua/treesitter.c

```diff
@@ -1422,15 +1424,33 @@ static int query_next_capture(lua_State *L)
     push_node(L, capture.node, lua_upvalueindex(2));  // [index, node]
 
     // Now check if we need to run the predicates
+    // a predicate is executed only once within the same match.
     uint32_t n_pred;
     ts_query_predicates_for_pattern(query, match.pattern_index, &n_pred);
 
-    if (n_pred > 0 && (ud->max_match_id < (int)match.id)) {
+    if (n_pred == 0) {
```

Comment:
I don't think we can do this now as we need the match table in `iter_captures` for the highlighter.

Reply:
I don't think this PR is not ready for review and is fixing the issue appropriately yet. I only rebased to the latest master fixing lots of conflicts. Design choices also need some changes (e.g. `match` must be always non-nil) now that we have #27132 and #27901.